File: rightView.py
# Time Complexity : O(N) 
# Space Complexity :    O(W) (max no.of leaves)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO

#bfs

class Solution:
    def __init__(self):
        self.res =[]
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root == None:
            return None
        if root == []:
            return []
        
        q = deque()
        q.append(root)
        
        while(len(q)!=0):
            cursize = len(q)
            
            for i in range(cursize):
                node = q.popleft()
                if node.left!=None:
                    q.append(node.left)
                if node.right!=None:
                    q.append(node.right)
                
                if i==cursize-1:
                    self.res.append(node.val)
        return self.res

# Time Complexity : O(N) 
# Space Complexity :    O(H)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO

#dfs
class Solution:
    def __init__(self):
        self.res =[]
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root==None:
            return None
        
        self.dfs(root,0)
        return self.res
    
    def dfs(self,root,level):
        if root==None:
            return 
        
        if len(self.res)==level:
            self.res.append(root.val)
        
        self.dfs(root.right,level+1)
        self.dfs(root.left,level+1)

        

        

File: rightview.java
// Time Complexity : O(n) all nodes
// Space Complexity : O(n) n/2 leaf nodes
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach:
// we are performing bfs and maintaing a list if the element is last node of the level we add it to the list.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) return result;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty())
        {
            int size = q.size();
            for(int i=0;i<size;i++)
            {
                TreeNode curr = q.poll();
                if(i==size-1)
                {
                    result.add(curr.val);
                }
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            } 
        }
        return result;
    }
}